<a href="https://colab.research.google.com/github/hiralverma5560773/hiralverma5560773-DataScience-GenAI-Submissions/blob/main/Assignment%206_02_DNN_101.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![](https://drive.google.com/uc?export=view&id=1xqQczl0FG-qtNA2_WQYuWePW9oU8irqJ)

# 6.02 Dense Neural Network (with PyTorch)
This will expand on our logistic regression example and take us through building our first neural network. If you haven't already, be sure to check (and if neccessary) switch to GPU processing by clicking Runtime > Change runtime type and selecting GPU. We can test this has worked with the following code:

In [ ]:
import torch

# Check for GPU availability
print("Num GPUs Available: ", torch.cuda.device_count())

Num GPUs Available:  1


Hopefully your code shows you have 1 GPU available! Next let's get some data. We'll start with another in-built dataset:

In [ ]:
# upload an in-built Python (OK semi-in-built) dataset
from sklearn.datasets import load_diabetes

import pandas as pd
import numpy as np

# import the data
data = load_diabetes()
data

{'data': array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
          0.01990749, -0.01764613],
        [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
         -0.06833155, -0.09220405],
        [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
          0.00286131, -0.02593034],
        ...,
        [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
         -0.04688253,  0.01549073],
        [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
          0.04452873, -0.02593034],
        [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
         -0.00422151,  0.00306441]]),
 'target': array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
         69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
         68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
         87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
        259.,  53., 190., 142.,  75., 142., 155., 225.,  59

We are working on a regression problem, with "structured" data which has already been cleaned and normalised. We can skip the usual cleaning/engineering steps. However, we do need to get the data into PyTorch:

In [ ]:
# Convert data to PyTorch tensors
X = torch.tensor(data.data, dtype=torch.float32)
y = torch.tensor(data.target, dtype=torch.float32).reshape(-1, 1) # Reshape y to be a column vector

Now our data is stored in tensors we can do train/test splitting as before (in fact we can use sklearn as before):

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

torch.Size([353, 10]) torch.Size([353, 1])
torch.Size([89, 10]) torch.Size([89, 1])


Now we can set up our batches for training. As we have a nice round 400 let's go with batches of 50 (8 batches in total). We'll also seperate the features and labels:

In [ ]:
from torch.utils.data import TensorDataset, DataLoader

# Create TensorDatasets and DataLoaders
train_dataset = TensorDataset(X_train, y_train)
train_loader = DataLoader(train_dataset, batch_size=50, shuffle=True)

test_dataset = TensorDataset(X_test, y_test)
test_loader = DataLoader(test_dataset, batch_size=50, shuffle=False)

Now its time to build our model. We'll keep it simple ... a model with an input layer of 10 features and then 2x _Dense_ (fully connected) layers each with 5 neurons and ReLU activation. Our output layer will be size=1 given this is a regression problem and we want a single value output per prediction.

This will be easier to understand if you have read through the logistic regression tutorial.

In [ ]:
import torch
import torch.nn as nn

# Define the model
class DiabetesModel(nn.Module):
    def __init__(self):
        super(DiabetesModel, self).__init__()
        # we'll set up the layers as a sequence using nn.Sequential
        self.layers = nn.Sequential(

            # first layer will be a linear layer that has 5x neurons
            # (5x sets of linear regression)
            # the layer takes the 10 features as input (i.e. 10, 5)
            nn.Linear(10, 5),

            nn.ReLU(), # ReLU activation

            # second linear layer again has 5 neurons
            # this time taking the input as the output of the last layer
            # (which had 5x neurons)
            nn.Linear(5, 5),

            nn.ReLU(), # ReLU again

            # last linear layer takes the output from the previous 5 neurons
            # this time its a single output with no activation
            # i.e. this is the predicitons (regression)
            nn.Linear(5, 1)
        )

    def forward(self, x):
        return self.layers(x) # pass the data through the layers

As before we need to create a model object, specify the loss (criterion) and an optimiser (which we cover next week):

In [ ]:
import torch.optim as optim

# Initialize the model, loss function, and optimizer
model = DiabetesModel()
criterion = nn.MSELoss() # MSE loss function
optimiser = optim.Adam(model.parameters(), lr=0.001)

Now we can train the model. Again, the logistic regression tutorial (6.01) may help you undertstand this:

In [ ]:
# Move model to GPU if available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Training loop (example - you'll likely want to add more epochs)
epochs = 100 # 100 epochs

for epoch in range(epochs):
  # use the train_loader to pass the inputs (x) and targets (y)
  for inputs, targets in train_loader:
    # pass to the GPU (hopefully)
    inputs, targets = inputs.to(device), targets.to(device)

    # pass model to GPU as well
    model.to(device)

    model.train() # put the model object in train mode
    optimiser.zero_grad() # reset the gradiants
    outputs = model(inputs) # create outputs
    loss = criterion(outputs, targets) # compare with Y to get loss
    loss.backward() # backpropogate the loss (next week)
    optimiser.step() # # update the parameters based on this round of training

  # every 10 steps we will print out the current loss
    if (epoch+1) % 10 == 0: # modular arithmetic
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {round(loss.item(), 4)}')

Epoch [10/100], Loss: 32218.1895
Epoch [10/100], Loss: 28822.1152
Epoch [10/100], Loss: 26224.6035
Epoch [10/100], Loss: 30192.6992
Epoch [10/100], Loss: 31181.7969
Epoch [10/100], Loss: 24244.875
Epoch [10/100], Loss: 36415.2695
Epoch [10/100], Loss: 13110.5186
Epoch [20/100], Loss: 28937.7344
Epoch [20/100], Loss: 31705.8945
Epoch [20/100], Loss: 22570.6504
Epoch [20/100], Loss: 33892.7031
Epoch [20/100], Loss: 26889.1367
Epoch [20/100], Loss: 32357.4434
Epoch [20/100], Loss: 30379.334
Epoch [20/100], Loss: 43714.8555
Epoch [30/100], Loss: 20940.0664
Epoch [30/100], Loss: 27930.6738
Epoch [30/100], Loss: 35799.4766
Epoch [30/100], Loss: 24780.4824
Epoch [30/100], Loss: 36075.3242
Epoch [30/100], Loss: 32395.8438
Epoch [30/100], Loss: 27459.9648
Epoch [30/100], Loss: 32824.293
Epoch [40/100], Loss: 27575.209
Epoch [40/100], Loss: 27031.791
Epoch [40/100], Loss: 30289.8223
Epoch [40/100], Loss: 28799.5488
Epoch [40/100], Loss: 29626.1895
Epoch [40/100], Loss: 28150.6914
Epoch [40/100],

We can see loss is significantly lower at the end than it was at the start. However, it is also bouncing around a little still which suggests the model needs more training (100 epochs is not a lot in deep learning terms). However, let's evaluate as before:

In [ ]:
# Evaluation (example)
model.eval() # testing mode
mse_values = [] # collect the MSE scores

with torch.no_grad():
    for inputs, targets in test_loader:
        inputs, targets = inputs.to(device), targets.to(device)
        outputs = model(inputs) # predict the test data

        # Calculate Mean Squared Error
        mse = criterion(outputs, targets) # calcualte mse for the batch
        mse_values.append(mse.item()) # add to the list of MSE values

# Calculate and print the average MSE
avg_mse = np.mean(mse_values)
print(f"Average MSE on test set: {avg_mse}")

Average MSE on test set: 14418.6728515625


MSE looks expected given training (no obvious sign of overfitting). However, we probably can get better results with tuning and more epochs.

Let's run the loop again a little differently to collect the predicted values (y_hat) and actuals (y) and add them to a dataset for comparions:

In [ ]:
# Evaluation
model.eval()
predictions = []
actuals = []

with torch.no_grad():
    for inputs, targets in test_loader:
        inputs, targets = inputs.to(device), targets.to(device)
        outputs = model(inputs)
        predictions.extend(outputs.cpu().numpy())
        actuals.extend(targets.cpu().numpy())

# Create DataFrame
results_df = pd.DataFrame({'Predicted': np.array(predictions).flatten(), 'Actual': np.array(actuals).flatten()})
results_df

Predicted  Actual
0   52.174736   219.0
1   46.873558    70.0
2   50.967339   202.0
3   64.950287   230.0
4   48.686230   111.0
..        ...     ...
84  41.975410   153.0
85  36.999256    98.0
86  32.317623    37.0
87  34.281322    63.0
88  42.830406   184.0

[89 rows x 2 columns]

Side-by-side, they don't look great. Can you improve them?

<br><br>

## EXERCISE #1
Try increasing the number of epochs to 1,000 (when the model is fairly well trained then the results printed for each 10x epochs will be fairly stable and not change much). Does this give better results?

<br><br>

## EXERCISE #2 (optional)
Try experimenting with the architecture (number of neurons and/or number of layers). Can we reach an optimal architecture?

In [ ]:
import torch
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Training loop (example - you'll likely want to add more epochs)
epochs = 1000 # 1000 epochs

for epoch in range(epochs):
  # use the train_loader to pass the inputs (x) and targets (y)
  for inputs, targets in train_loader:
    # pass to the GPU (hopefully)
    inputs, targets = inputs.to(device), targets.to(device)

    # pass model to GPU as well
    model.to(device)

    model.train() # put the model object in train mode
    optimiser.zero_grad() # reset the gradiants
    outputs = model(inputs) # create outputs
    loss = criterion(outputs, targets) # compare with Y to get loss
    loss.backward() # backpropogate the loss (next week)
    optimiser.step() # # update the parameters based on this round of training

  # every 100 steps we will print out the current loss
    if (epoch+1) % 100 == 0: # modular arithmetic
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {round(loss.item(), 4)}')

Epoch [100/1000], Loss: 4799.9087
Epoch [100/1000], Loss: 4719.3877
Epoch [100/1000], Loss: 2616.5898
Epoch [100/1000], Loss: 4722.2944
Epoch [100/1000], Loss: 3783.5425
Epoch [100/1000], Loss: 3569.241
Epoch [100/1000], Loss: 4267.1528
Epoch [100/1000], Loss: 12200.2178
Epoch [200/1000], Loss: 3736.2693
Epoch [200/1000], Loss: 3700.7468
Epoch [200/1000], Loss: 4460.7012
Epoch [200/1000], Loss: 3833.9832
Epoch [200/1000], Loss: 3862.093
Epoch [200/1000], Loss: 2761.3777
Epoch [200/1000], Loss: 3760.8809
Epoch [200/1000], Loss: 1637.4178
Epoch [300/1000], Loss: 3657.4644
Epoch [300/1000], Loss: 3507.1411
Epoch [300/1000], Loss: 3059.6829
Epoch [300/1000], Loss: 3957.6006
Epoch [300/1000], Loss: 3185.6194
Epoch [300/1000], Loss: 3357.842
Epoch [300/1000], Loss: 3544.7749
Epoch [300/1000], Loss: 2304.5669
Epoch [400/1000], Loss: 4037.0874
Epoch [400/1000], Loss: 3356.8555
Epoch [400/1000], Loss: 2057.363
Epoch [400/1000], Loss: 3493.7891
Epoch [400/1000], Loss: 3569.8286
Epoch [400/1000],

In [ ]:
model.eval() # Set the model to evaluation mode
mse_values = [] # List to store MSE for each batch

with torch.no_grad(): # Disable gradient calculations during evaluation
    for inputs, targets in test_loader:
        inputs, targets = inputs.to(device), targets.to(device)
        outputs = model(inputs) # Get model predictions

        # Calculate Mean Squared Error for the current batch
        mse = criterion(outputs, targets)
        mse_values.append(mse.item())

# Calculate and print the average MSE across all test batches
avg_mse_new = np.mean(mse_values)
print(f"Average MSE on test set after 1000 epochs: {avg_mse_new}")

Average MSE on test set after 1000 epochs: 2840.306884765625


In [ ]:
model.eval() # Set the model to evaluation mode
predictions = []
actuals = []

with torch.no_grad(): # Disable gradient calculations
    for inputs, targets in test_loader:
        inputs, targets = inputs.to(device), targets.to(device)
        outputs = model(inputs) # Get predictions
        predictions.extend(outputs.cpu().numpy())
        actuals.extend(targets.cpu().numpy())

# Create a DataFrame to compare predicted and actual values
results_df_new = pd.DataFrame({'Predicted': np.array(predictions).flatten(), 'Actual': np.array(actuals).flatten()})
display(results_df_new)

Predicted  Actual
0   143.703491   219.0
1   176.980194    70.0
2   139.807510   202.0
3   294.964661   230.0
4   124.107155   111.0
..         ...     ...
84  115.743660   153.0
85   87.155128    98.0
86   78.253540    37.0
87   64.716560    63.0
88  156.425812   184.0

[89 rows x 2 columns]

Increasing the number of epochs from 100 to 1000 resulted in a significant reduction in the Mean Squared Error  on the test set, from approximately 14418.67 to 2840.31. The model learned more effectively with additional training iterations. The predictions after 1000 epochs are much more accurate and align better with the actual target values. The model was indeed under-trained after only 100 epochs.